src/app/content/highlights/utils/paginationUtils.ts

```diff
@@ -0,0 +1,89 @@
+import add from 'lodash/fp/add';
+import pickBy from 'lodash/fp/pickBy';
+import { reduceUntil } from '../../../fpUtils';
+import { isDefined } from '../../../guards';
+import { ArchiveTree, LinkedArchiveTreeSection } from '../../types';
+import {
+  archiveTreeContainsNode,
+  archiveTreeSectionIsChapter,
+  findArchiveTreeNode,
+  findTreePages
+} from '../../utils/archiveTreeUtils';
+import { stripIdVersion } from '../../utils/idUtils';
+import { CountsPerSource } from '../types';
+
+const totalOfCountsPerSource = (perSource: CountsPerSource) => Object.values(perSource).reduce(add, 0);
+
+const diffSourceCounts = (base: CountsPerSource, diff: CountsPerSource) => Object.keys(base).reduce(
+  (counts: CountsPerSource, key: string) => ({...counts, [key]: base[key] - (diff[key] || 0)}),
+  {} as CountsPerSource
+);
+
+export const getRemainingSourceCounts = (
+  loadedCounts: CountsPerSource,
+  totalCounts: CountsPerSource
+): CountsPerSource => {
+  return diffSourceCounts(totalCounts, loadedCounts);
+};
+
+/*
+ * in order to avoid passing all page ids for all selected chapters in
+ * api requests to get highlights, we calculate how many ids are necessary
+ * to fill a response based on the given CountsPerSource
+ */
+export const getNextPageSources = (
+  remainingCounts: CountsPerSource,
+  tree: ArchiveTree,
+  nextPageSize: number
```

Comment:
remainingCounts comes from selectors.remainingSourceCounts, tree comes from book.tree, book is required to be known when getting highlights because it is used as the scopeId. nextPageSize can be anything but should be the same as the perPage passed to getHighlights